rllib/core/rl_module/marl_module.py

```diff
@@ -28,22 +43,122 @@ class MultiAgentRLModule(RLModule):
+        `observation_space`, `action_space`, or `model_config`, you can specify these
+        keys at the top level of the config, and the module spec will inherit the
+        values from the top level config.
+
+        Examples:
+
+        .. code-block:: python
+
+            config = {
+                "modules": {
+                    "module_1": {
+                        "module_class": "RLModule1",
+                        "observation_space": gym.spaces.Box(...),
+                        "action_space": gym.spaces.Discrete(...),
+                        "model_config": {hidden_dim: 256}
+                    },
+                    "module_2": {
+                        "module_class": "RLModule2",
+                        "observation_space": gym.spaces.Box(...),
+                    }
+                },
+                "action_space": gym.spaces.Box(...),
+                "model_config": {hidden_dim: 32}
+            }
+
+            # This is equivalent to the following config:
+
+            config = {
+                "modules": {
+                    "module_1": {
+                        "module_class": "RLModule1",
+                        "observation_space": gym.spaces.Box(...),
+                        "action_space": gym.spaces.Discrete(...),
+                        "model_config": {hidden_dim: 256}
+                    },
+                    "module_2": {
+                        "module_class": "RLModule2",
+                        "observation_space": gym.spaces.Box(...),
+                        "action_space": gym.spaces.Box(...), # Inherited
+                        "model_config": {hidden_dim: 32} # Inherited
+                    }
+                },
+            }
+
+        Args:
+            config: A config dict that contains the module configs. See above for the
+                format required.
+
+            Returns:
+                The MultiAgentRLModule.
         """
 
+        module_configs: Dict[ModuleID, Any] = _get_module_configs(config)
+        cls.__check_module_configs(module_configs)
+
+        multiagent_module = cls()
+
+        for module_id, module_spec in module_configs.items():
+            module_cls: RLModule = module_spec.pop("module_class")
+            module = module_cls.from_model_config(**module_spec)
+            multiagent_module.add_module(module_id, module)
+
+        return multiagent_module
+
+    @classmethod
+    def __check_module_configs(cls, module_configs: Dict[ModuleID, Any]):
+        """Checks the module configs for validity.
+
+        The module_configs be a mapping from module_ids to a dict that contains the
+        following required keys: `module_class`, `observation_space`, `action_space`,
+        `model_config`.
+
+        Args:
+            module_configs: The module configs to check.
+
+        Raises:
+            ValueError: If the module configs are invalid.
+        """
+        REQUIRED_KEYS = {
```

Comment:
Please add a TODO that any still-required config dicts will soon be replaced with proper config objects (for the respective Module classes).

Reply:
This is module spec requirement of this module itself. What you are saying is on the algorithm_config level. I don't want any information leakage from this module to how the caller uses it. Whether it's a dict or a custom dataclass when it comes to this module it should be a simple dict. We can change it later down the road if it become struggling to go back and forth between different formats.